### ISSUE TITLE
Initial FCI result of ZCASSCF depends on the "algorithm" keyword

### ISSUE THREAD

**Issue by stknecht** [nubakery/bagel#147]
I am running some simple tests for the HF molecule with a minimal CAS(2,2) using the relativistic CASSCF (ZCASSCF) code. In the initial FCI step I noticed that depending on the string (`second` or `noopt`) given to the input option `"algorithm": string`, I get different FCI results. In the case of `noopt` the result agrees with a corresponding ZFCI calculation (and a FCI calculation performed in DIRAC). Based on the comparisons, I am guessing that the FCI start guess in the case of  `second` might be wrong. 

For `noopt` I get 

```
  === Relativistic FCI iteration ===

                 * sigma vector                                0.00
                 * davidson                                    0.00
                 * error                                       0.00
                 * denominator                                 0.00
      0   0        -99.54010003     7.46e-02      0.00
                 * sigma vector                                0.00
                 * davidson                                    0.00
                 * error                                       0.00
                 * denominator                                 0.00
      1   0        -99.54764068     1.39e-04      0.00
                 * sigma vector                                0.00
                 * davidson                                    0.00
                 * error                                       0.00
                 * denominator                                 0.00
      2   0        -99.54764073     1.85e-06      0.00
                 * sigma vector                                0.00
                 * davidson                                    0.00
                 * error                                       0.00
                 * denominator                                 0.00
      3   0  *     -99.54764073     1.13e-15      0.00

     * ci vector, state   0

     * ci vector, state   0
       2.  (-0.9991487092,0.0000000000)
       .2  (0.0412533228,-0.0000000000)

     * ci vector, state   0
```

whereas for `second` I get in the initial step

```
  === Relativistic FCI iteration ===

                 * sigma vector                                0.00
                 * davidson                                    0.00
                 * error                                       0.00
                 * denominator                                 0.00
      0   0        -99.54010003     1.89e-02      0.00
                 * sigma vector                                0.00
                 * davidson                                    0.00
                 * error                                       0.00
                 * denominator                                 0.00
      1   0  *     -99.54061559     4.87e-14      0.00

     * ci vector, state   0

     * ci vector, state   0
       2.  (-0.9999377267,0.0000000000)
       .2  (0.0111598666,-0.0000000000)

     * ci vector, state   0
``` 

For reference below the input for `noopt`, for `second` just change the `algorithm` keyword:
```
{ "bagel" : [ 
{
  "finite_nucleus" : true,
  "title" : "molecule",
  "basis" : "3-21g",
  "df_basis" : "tzvpp-jkfit",
  "angstrom" : false,
  "geometry" : [ 
    { "atom" : "F",  "xyz" : [    0.000000,    0.000000,     -1.3721434143512670]},
    { "atom" : "H",  "xyz"  : [   0.000000,    0.000000,      0.3286970219154437]}
  ]
},
{
  "title"  : "zcasscf",
  "algorithm" : "noopt",
  "state" : [1],
  "thresh" : 5.0e-7,
  "thresh_micro" : 1.0e-8,
  "thresh_fci" : 1.0e-9,
  "natocc" : true,
  "print_thresh": 1.0e-2,
  "only_electrons" : true,
  "active": [5,9],
  "nact"  : 2,
  "nclosed"  : 4,
  "pop" : true,
  "maxiter"  : 100,
  "maxiter_scf"  : 100,
  "maxiter_fci"  : 100,
  "maxiter_micro" : 20, 
  "diis_start"   : 5 
}
]}
```

I compiled the BAGEL version `commit bfceffea5725992c708a9ae03f26604e2fcc1b15` with GNU 6.5, MKL 19.1 and OpenMPI on a Mac with macOS Mojave.

**Comment by shiozaki** [comment]
It is not wrong. ZCASSCF is Kramer’s adapted but DHF is not, so it tries to
make orbitals that are Kramer’s adapted in s crude way. In practice it
doesn’t matter much.

**Comment by shiozaki** [comment]
To observe this, you could perhaps run a calculation with tiny magnetic field 0.000001 T, which enforces CASSCF to be run without Kramers symmetry (in this case, second should match with noopt). I assume that second without field and with field converge to the same solution (unless you have a highly symmetric molecules and active spaces are not identical).

**Comment by stknecht** [comment]
Hi Toru,

thanks for your quick answer. I am not saying that ZCASSCF is wrong. What I am saying is that in the initial FCI step (prior to any orbital optimization) ZCASSCF gives different results depending on whether ‘noopt’ or ‘second’ is specified for the algorithm keyword. shouldn’t it be the same since in either case we start from the same DHF orbitals and perform a FCI calculation within the specified CAS?

ZCASSCF with noopt yields the same as ZFCI which is what I would have expected and so should ZCASSCF in the 0-th macro iteration (before we start orbital optimization) but that’s not what I find.

**Comment by stknecht** [comment]
@shiozaki 
I tried your suggestion with the magnetic field but the result is the same as without the magnetic field. The initial FCI step of ZCASSCF which provides the **reference CI vector** prior to the first macro iteration gives still different results depending on the `algorithm` keyword. IMHO, that does NOT make any sense. The FCI energy in 0-th macro iteration should always be the same, shouldn't it?
What I get is:

- `noopt`: -99.54764073
- `second`: -99.54061559

I don't see any reason why the two cases should be any different, in either case we start from the same DHF orbitals which match perfectly (population analysis) with the DHF orbitals calculated with DIRAC. The FCI energy of DIRAC agrees also nicely with the FCI energy of the `noopt` case. 
Maybe it's obvious for you what I am doing wrong?

**Comment by shiozaki** [comment]
No orbital sub spaces are missed up in symmetry purification process (I
think it’s diagonalizing fock there), so it is working as intended. Perhaps
the right things to do is to implement adapted HF but I have no time.

**Comment by stknecht** [comment]
I don't understand your explanation, maybe I am just too stupid to see it. 
What I am saying is that for my small example the (unrestricted) DHF orbitals from BAGEL seem to match perfectly the restricted DHF orbitals from DIRAC and so does the 0-th macro iteration FCI energy **IF** I use the `noopt` option. If I use `second` within the very same ZCASSCF input I get a different FCI energy in the 0-th macro iteration. Why? Shouldn't they be the same? 
Does the program do anything special (with the input orbitals) prior to calculating the FCI energy in the 0-th macro iteration if `algorithm == second` which is not done if `algorithm == noopt`? 

I really don't think unrestricted DHF vs restricted DHF orbitals are the issue here. The system is the H-F molecule (closed-shell, 10 electrons) with a minimal basis set.

**Comment by shiozaki** [comment]
It is the issue. No opt doesn’t do symmetrization and fci itself doesn’t
assume symmetry. If you want to discuss please write to me directly and we
can set up a Skype.
